common/persistence/elasticsearch/esProcessor.go

```diff
@@ -247,81 +283,96 @@ func (p *esProcessorImpl) sendToAckChan(visibilityTaskKey string, ack bool) {
 	})
 }
 
-func (p *esProcessorImpl) getVisibilityTaskKey(request elastic.BulkableRequest) string {
+func (p *esProcessorImpl) extractVisibilityTaskKey(request elastic.BulkableRequest) string {
 	req, err := request.Source()
 	if err != nil {
-		p.logger.Error("Get request source err.", tag.Error(err), tag.ESRequest(request.String()))
+		p.logger.Error("Unable to get ES request source.", tag.Error(err), tag.ESRequest(request.String()))
 		p.metricsClient.IncCounter(metrics.ElasticSearchVisibility, metrics.ESBulkProcessorCorruptedData)
 		return ""
 	}
 
-	var key string
 	if len(req) == 2 { // index or update requests
 		var body map[string]interface{}
-		if err := json.Unmarshal([]byte(req[1]), &body); err != nil {
-			p.logger.Error("Unmarshal index request body err.", tag.Error(err))
+		if err = json.Unmarshal([]byte(req[1]), &body); err != nil {
+			p.logger.Error("Unable to unmarshal ES request body.", tag.Error(err))
 			p.metricsClient.IncCounter(metrics.ElasticSearchVisibility, metrics.ESBulkProcessorCorruptedData)
 			return ""
 		}
 
 		k, ok := body[definition.VisibilityTaskKey]
 		if !ok {
-			// must be bug in code and bad deployment, check processor that add es requests
-			panic("VisibilityTaskKey not found")
-		}
-		key, ok = k.(string)
-		if !ok {
-			// must be bug in code and bad deployment, check processor that add es requests
-			panic("VisibilityTaskKey is not string")
-		}
-	} else { // delete requests
-		var body map[string]map[string]interface{}
-		if err := json.Unmarshal([]byte(req[0]), &body); err != nil {
-			p.logger.Error("Unmarshal delete request body err.", tag.Error(err))
+			p.logger.Error("Unable to extract VisibilityTaskKey from ES request.", tag.ESRequest(request.String()))
```

Comment:
so not logger.Fatal() here? Don't want to panic?

Reply:
I think these `panic` are because of backward compatibility. I can't imagine case where this happen and yes, I prefer to log error/metric instead of panic.